services/src/main/java/org/openlearn/service/StorageService.java

```diff
@@ -109,23 +144,101 @@ public InputStream getUpload(Long assignmentId, Long portfolioId, String fileNam
+			s3client.deleteObject(new DeleteObjectRequest(bucket, key));
+			fileRepository.delete(fileInformation);
 		} catch(Exception e) {
 			log.error(e.getMessage());
 		}
 	}
 
-	private String getS3FilePrefix(Long assignmentId, Long portfolioId) {
+	public void deleteUploads(List<FileInformation> files) {
+		AmazonS3 s3client = AmazonS3ClientBuilder.defaultClient();
+		List<String> buckets = Lists.transform(files, this::retrieveBucket);
+		List<String> keys = Lists.transform(files, this::retrieveKeyName);
+
+		Map<String, List<String>> bucketToKeys = new HashMap<>();
+
+		// Test if there's only one bucket (which will be the typical case)
+		Set<String> uniqueBuckets = new HashSet<>(buckets);
+		if (uniqueBuckets.size() > 1) {
+			bucketToKeys.put(buckets.get(0), keys);
+		} else { // There are multiple buckets in the given file set
+
+			for (int index = 0; index < buckets.size(); index++) {
+				String bucket = buckets.get(index);
+				String key = keys.get(index);
+				if (bucketToKeys.containsKey(bucket)) {
+					bucketToKeys.get(bucket).add(key);
+				} else {
+					List<String> list = new ArrayList<>();
+					list.add(key);
+					bucketToKeys.put(bucket, list);
+				}
+			}
+		}
+
+		bucketToKeys.forEach((bucket, keys1) -> {
+            DeleteObjectsRequest request = new DeleteObjectsRequest(bucket);
+            List<DeleteObjectsRequest.KeyVersion> keyVersions = new ArrayList<>(keys1.size());
+            for (String key : keys1)
+                keyVersions.add(new DeleteObjectsRequest.KeyVersion(key));
+            request.setKeys(keyVersions);
+            s3client.deleteObjects(request);
+        });
+	}
+
+	private String createS3FilePrefix(Long assignmentId, Long portfolioId, Long uploadedByUserId) {
 		// s3 file name is a_[assignmentid]_[filename] for assignments and p_[portfolioid]_[filename] for portfolios
-		String assignmentStr = assignmentId != null ? "a_" + assignmentId.toString() + "_" : "";
-		String portfolioStr = portfolioId != null ? "p_" + portfolioId.toString() + "_" : "";
-		String prefix = assignmentStr + portfolioStr;
+		// and a unique GUID hash.
+		String assignmentStr = assignmentId != null ? "a_" + assignmentId.toString() + "/" : "";
+		String portfolioStr = portfolioId != null ? "p_" + portfolioId.toString() + "/" : "";
+		String userStr = uploadedByUserId.toString();
+		String prefix = assignmentStr + portfolioStr + userStr + "/";
 		return prefix;
 	}
 
+	private String retrieveBucket(FileInformation fileInformation) {
+		String bucketAndKey = retrieveFullPath(fileInformation);
+		int separatingIndex = bucketAndKey.indexOf("/");
+		return bucketAndKey.substring(0, separatingIndex);
+	}
+
+	private String retrieveKeyName(FileInformation fileInformation) {
+		String bucketAndKey = retrieveFullPath(fileInformation);
+		int separatingIndex = bucketAndKey.indexOf("/");
+		return bucketAndKey.substring(separatingIndex+1);
+	}
+
+	private String retrieveFullPath(FileInformation fileInformation) {
```

Comment:
We could, but Normalization purist would claim that the bucket and key are "calculable" from the file url. I'm also of a mind that storing the bucket and key would be ideal, the only problem is the handling of existing data. How do migrate existing schema and data correctly? I don't think that's too big of an issue (since it's just our development environments), but what about the staging environment we've been deploying to? Don't know if that DB is wiped every time...